src/gencertchain.c

```diff
@@ -0,0 +1,321 @@
+  FILE* file = fopen(filepath, "rb");
+  if (file == NULL) {
+    fprintf(stderr, "fopen %s: %s\n", filepath, strerror(errno));
+    exit(EXIT_FAILURE);
+  }
+  if (fseek(file, 0, SEEK_END) != 0) {
+    fprintf(stderr, "fseek %s: %s\n", filepath, strerror(errno));
+    exit(EXIT_FAILURE);
+  }
+  const long filesize = ftell(file);
+  if (filesize == -1) {
+    fprintf(stderr, "ftell %s: %s\n", filepath, strerror(errno));
+    exit(EXIT_FAILURE);
+  }
+  rewind(file);
+  sxg_buffer_resize(filesize, dst);
+  int read_size = fread(dst->data, sizeof(uint8_t), filesize, file);
+  if (read_size != filesize) {
+    fprintf(stderr, "fread %s: %s\n", filepath, strerror(errno));
+    exit(EXIT_FAILURE);
+  }
+  fclose(file);
+}
+
+static void serialize_ocsp_file(const char* ocsp_path, OCSP_RESPONSE** dst) {
+  sxg_buffer_t ocsp_content = sxg_empty_buffer();
+  load_file(ocsp_path, &ocsp_content);
+  const uint8_t* ocsp_ptr = ocsp_content.data;
+  d2i_OCSP_RESPONSE(dst, &ocsp_ptr, ocsp_content.size);
+  if (dst == NULL) {
+    fprintf(stderr, "Failed to parse OCSP response: %s\n", ocsp_path);
+    exit(EXIT_FAILURE);
+  }
+}
+
+static void make_glob_pattern(const char* base_path, sxg_buffer_t* dst) {
+  sxg_buffer_release(dst);
+  if (!sxg_write_string(base_path, dst) ||
+      !sxg_write_string("/*.sct", dst) ||
+      !sxg_write_byte('\0', dst)) {
+    fprintf(stderr, "Failed to make glob pattern.\n");
+    exit(EXIT_FAILURE);
+  }
+}
+
+static void serialize_sct_files(const char* sct_path, sxg_buffer_t* sct_list) {
+  glob_t glob_result;
+
+  // Load all sct files at sct_path into sxg_buffer_t array.
+  sxg_buffer_t glob_pattern = sxg_empty_buffer();
+  make_glob_pattern(sct_path, &glob_pattern);
+  if (glob((const char*)glob_pattern.data, GLOB_ERR | GLOB_NOESCAPE,
+           NULL, &glob_result) != 0) {
+    globfree(&glob_result);
+    fprintf(stderr, "Failed to glob at: %s\n", sct_path);
+    exit(EXIT_FAILURE);
+  }
+  sxg_buffer_release(&glob_pattern);
+  const size_t files = glob_result.gl_pathc;
+  sxg_buffer_t* buffers = OPENSSL_malloc(sizeof(sxg_buffer_t) * glob_result.gl_pathc);
+  for (size_t i = 0; i < files; ++i) {
+    buffers[i] = sxg_empty_buffer();
+    load_file(glob_result.gl_pathv[i], &buffers[i]);
+    if (buffers[i].size > USHRT_MAX) {
+      fprintf(stderr, "Too long sct file: %s\n", glob_result.gl_pathv[i]);
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  // Accumulate total size.
+  size_t total_size = 0;
+  for (size_t i = 0; i < files; ++i) {
+    total_size += buffers[i].size;
+    if (total_size > USHRT_MAX) {
+      fprintf(stderr, "Too long total length of *.sct files: %s\n", sct_path);
+      exit(EXIT_FAILURE);
+    }
+  }
+  total_size += files * 2;  // 16-bits length prefix on each sct.
+
+  // Serialize sct payloads.
+  bool success = sxg_write_int(total_size, 2, sct_list);
+  for (size_t i = 0; i < files && success; ++i) {
+    success = success && sxg_write_int(buffers[i].size, 2, sct_list) &&
```

Comment:
Is the `success &&` required since the loop condition also checks this?

Reply:
removed.